communication/src/AsyncTlsConnection.cpp

```diff
@@ -99,81 +108,57 @@ void AsyncTlsConnection::readMsg() {
                  return dispose();
                }
 
-               // This is a bug in boost::asio (boost 1.65) that was fixed in later versions. `async_read` is
-               // only supposed to return a full buffer or an error. However, this bug allows, no error but a
-               // return of exactly zero bytes read. In this case, we don't know the state of the
-               // stream, so it's safest to dispose of the connection.
-               // https://github.com/boostorg/asio/pull/182
-               // https://github.com/boostorg/asio/commit/ce7e3bbf4b7070b8292df50d3514c34ce0353684
-               // https://github.com/chriskohlhoff/asio/commit/57b2ef19b013dd1fd8660af28398d3d332d1ea97
-               if (bytes_transferred == 0) {
-                 LOG_ERROR(logger_, "Short read (0 bytes) of message occurred" << KVLOG(peer_id_.value(), msg_size));
-                 return dispose();
-               }
-
-               // This would be a much more severe bug that is undocumented, and also that we have
-               // never seen. This check is here just in case, because we do not want to revert our
-               // composed operation (async_read) to partial read (async_read_some) methods and make
-               // things more complex.
-               if (bytes_transferred != msg_size) {
-                 LOG_ERROR(logger_,
-                           "ASIO violated the contract for async_read and did not fill the buffer. Closing connection."
-                               << KVLOG(peer_id_.value(), msg_size, bytes_transferred));
-                 return dispose();
-               }
-
                // The Read succeeded.
-               boost::system::error_code _ec;
                LOG_DEBUG(logger_, "Cancelling read timer: " << KVLOG(peer_id_.value(), (void*)read_msg_.data()));
-               read_timer_.cancel(_ec);
-               tlsTcpImpl_.histograms_.received_msg_size->record(bytes_transferred);
+               read_timer_.cancel();
```

Comment:
It's a good point. The `error_code version` is deprecated though. Looking at the code, it still shows it's possible to throw if the underlying implementation returns an error code. I feel like it may not actually be possible to throw though, as why else would the error code version become deprecated?

Reply:
I talked it out with @f-squirrel  offline. We can't really handle this situation anything, so throwing seems to be the right option.